config-controller/internal/controller/policy_controller.go

```diff
@@ -0,0 +1,134 @@
+}
+
+//+kubebuilder:rbac:groups=config.stackrox.io,resources=policies,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=config.stackrox.io,resources=policies/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=config.stackrox.io,resources=policies/finalizers,verbs=update
+
+func (r *PolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	rlog := log.FromContext(ctx)
+	rlog.Info("Reconciling", "namespace", req.Namespace, "name", req.Name)
+
+	// Get the policy CR
+	policyCR := &configstackroxiov1alpha1.Policy{}
+	if err := r.Client.Get(ctx, req.NamespacedName, policyCR); err != nil {
+		if k8serr.IsNotFound(err) {
+			// Must have been deleted
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("Failed to get policy: namespace=%s, name=%s", req.Namespace, req.Name)
+	}
+
+	desiredState := &storage.Policy{}
+	if err := protojson.Unmarshal([]byte(policyCR.Spec.Policy), protoadapt.MessageV2Of(desiredState)); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "Failed to parse policy JSON")
+	}
+
+	// GET policy from Central
+	svc := v1.NewPolicyServiceClient(r.conn)
+	allPolicies, err := svc.ListPolicies(ctx, &v1.RawQuery{})
+
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "Failed to list all policies")
+	}
+
+	for _, policy := range allPolicies.Policies {
+		if policy.Name == req.Name {
+			desiredState.Id = policy.Id
+			if _, err = svc.PutPolicy(ctx, desiredState); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, fmt.Sprintf("Failed to PUT policy %s", req.Name))
+			}
+			policyCR.Status.Accepted = true
+			policyCR.Status.Message = "Successfully updated policy"
+			if err = r.Client.Status().Update(ctx, policyCR); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, fmt.Sprintf("Failed to set status on policy %s", req.Name))
+			}
+			return ctrl.Result{}, nil
+		}
+	}
+
+	if _, err = svc.PostPolicy(ctx, &v1.PostPolicyRequest{Policy: desiredState}); err != nil {
+		wrappedErr := errors.Wrap(err, fmt.Sprintf("Failed to create policy %s", req.Name))
+		policyCR.Status.Accepted = false
+		policyCR.Status.Message = wrappedErr.Error()
+		if err = r.Client.Status().Update(ctx, policyCR); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, fmt.Sprintf("Failed to set status on policy %s", req.Name))
+		}
+		return ctrl.Result{}, wrappedErr
+	} else {
+		policyCR.Status.Accepted = true
+		policyCR.Status.Message = "Successfully created policy"
+		if err = r.Client.Status().Update(ctx, policyCR); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, fmt.Sprintf("Failed to set status on policy %s", req.Name))
+		}
+	}
+
+	// Report status
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PolicyReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	conn, err := common.GetGRPCConnection(auth.TokenAuth(), logger.NewLogger(roxctlIO.DefaultIO(), printer.DefaultColorPrinter()))
```

Comment:
My understanding is the reconciler will be a service alongside central. If yes, then it should have its own log and we should add that log to the diag bundle.

The other question - if we control the service from being deployed or not based on feature flag, what happens once the feature is enabled? Do we always have this extra service even if folks don't use policy management via code?

Reply:
This gets at differing meanings of the term "Feature Flag". 

Boo's question seems to assume that flags are only to gate features under development until they are ready to be released.

I subscribe to a broader view that some version of a flag could be left available to turn a feature on or off as need. This controller is a good use-case for that type of flag.

Reply:
> it should have its own log and we should add that log to the diag bundle

Reply:
> it should have its own log and we should add that log to the diag bundle
This sounds correct, but I'm not 100% sure how simple this is to do. It feels like this interface is assumed to only be used by roxctl, but I'll look into it.

> what happens once the feature is enabled?
I was thinking the feature flag would be managed by a helm chart parameter.  So with that parameter set to true, this would enable the feature flag in central _and_ deploy the config-controller.

Reply:
I think we can defer the log question to a follow-up PR. And if we need to discuss the semantics of the feature flag (I'm not sure exactly where we landed on that conversation), perhaps we can follow up in Slack.

Reply:
Feature Flag & Configuration are distinct from each other.

Configuration: Can enable or disable a feature/component of a system.

Feature Flag: Controls if a feature is available, e.g. used to hide features from customers completely.

My 2 cents, otherwise I don't have hard feelings, depends what you want.

Reply:
tbh I would always enable the config-controller, it is a simple process. disabling it more a debug task to me.

Reply:
I think this could be considered a "stretch" feature.  It's not required to allow customers to disable the config-controller, but it is a nice-to-have if we have extra time.

Reply:
https://issues.redhat.com/browse/ROX-25859